src/jltypes.c

```diff
@@ -2458,34 +1139,71 @@ static jl_value_t *inst_tuple_w_(jl_value_t *t, jl_value_t **env, size_t n,
         }
     }
     jl_value_t *result = inst_datatype((jl_datatype_t*)tt, ip_heap, iparams, ntp, cacheable,
-                                       stack, env, n);
+                                       stack, env);
     JL_GC_POP();
     return result;
 }
 
-static jl_value_t *inst_type_w_(jl_value_t *t, jl_value_t **env, size_t n,
-                                jl_typestack_t *stack, int check)
+static jl_value_t *inst_type_w_(jl_value_t *t, jl_typeenv_t *env, jl_typestack_t *stack, int check)
 {
-    size_t i, j;
-    if (n == 0) return t;
+    size_t i;
     if (jl_is_typevar(t)) {
-        for(i=0; i < n; i++) {
-            if (env[i*2] == t) {
-                jl_value_t *val = env[i*2+1];
-                if (check && !jl_is_typevar(val) && !jl_subtype(val, t, 0)) {
-                    jl_type_error_rt("type parameter",
-                                     jl_symbol_name(((jl_tvar_t*)t)->name),
-                                     t, val);
-                }
+        jl_typeenv_t *e = env;
+        while (e != NULL) {
+            if (e->var == (jl_tvar_t*)t) {
+                jl_value_t *val = e->val;
+                // TODO jb/subtype this seems unnecessary
+                //if (check && !jl_is_typevar(val) && !within_typevar(val, (jl_tvar_t*)t)) {
+                //    jl_type_error_rt("type parameter",
+                //                     jl_symbol_name(((jl_tvar_t*)t)->name), t, val);
+                //}
                 return val;
             }
+            e = e->prev;
         }
         return (jl_value_t*)t;
     }
+    if (jl_is_unionall(t)) {
+        jl_typestack_t *sp = stack;
+        while (sp != NULL) {
+            if (sp->ua == t)
+                return sp->ua_new;
+            sp = sp->prev;
+        }
+        if (!jl_has_free_typevars(t))
+            return t;
+        jl_unionall_t *ua = (jl_unionall_t*)t;
+        jl_value_t *res=NULL, *lb=ua->var->lb, *ub=ua->var->ub;
+        JL_GC_PUSH3(&lb, &ub, &res);
+        res = jl_new_struct(jl_unionall_type, ua->var, NULL);
+        jl_typestack_t top = { {t}, res, stack };
+        if (jl_has_bound_typevars(ua->var->lb, env) || jl_has_bound_typevars(ua->var->ub, env)) {
```

Comment:
possibly better not to merge these two conditionals? they don't seem dependent on each other.